text/0000-trust-boundary-specification.md

```diff
@@ -0,0 +1,128 @@
+
+Examples of Pony operations that can have undefined results include unsafe maths operations, like unchecked division.
+
+### Undefined behaviour
+
+If an expression with undefined behaviour is evaluated, or if the program does a FFI call with undefined behaviour (as defined by the specification of the language the called function is written in), every behaviour in the program is free from the as-if rule, including the capability security guarantees. Conversely, violating the capability security guarantees in an unsafe operation or FFI call results in undefined behaviour.
+
+Examples of Pony operations that can have undefined behaviour include raw pointer operations (not available currently).
+
+### Out-of-memory errors
+
+When encountering an unrecoverable out-of-memory situation (e.g. a very deep recursion, an object allocation failure, etc.) the program is required to terminate and to print an explanatory error message to the user.
+
+### The new trust boundary
+
+The new trust boundary will be separated into multiple levels, with increasing permissiveness. Each level includes the permissions of the previous levels.
+
+- At the first level, a package is allowed to do operations that can have undefined results.
+- At the second level, a package is allowed to do operations that can have undefined behaviour.
+- At the third level, a package is allowed to do FFI calls.
+
+The main reason for separating undefined behaviour operations and the FFI into different levels is for general side-effect control. An FFI call can do _anything_, including writing to files or to the network, but a Pony operation with potential undefined behaviour has a limited scope.
+
+Out-of-memory errors aren't covered by the trust boundary because trying to do so would greatly limit the usefulness of untrusted packages.
+
+## User-facing changes
+
+### In the compiler
+
+With the trust boundary now being split into multiple levels, the `--safe` flag must also be separated into multiple levels. The new flags, `--safe-1`, `--safe-2`, and `--safe-3`, can be used in the same way as `--safe` to specify the trust level of packages in the program. `--safe` will be removed.
+
+- Higher-level flags take precedence over lower-level flags.
+- In the same way `--safe` works currently, if any of those flags is specified, any package not in the trusted list is treated as untrusted.
+- If none of those flags are specified, all packages are treated as trusted with a level of 3.
+- The `builtin` package is always treated as trusted with a level of 3.
+
+### In the language
+
+New syntax is needed to mark unsafe operations. We propose to use new annotations for this purpose: `unsafe-1`, `unsafe-2` and `unsafe-3`. These annotations map directly to the various levels of the trust boundary. These annotations are only allowed on function definitions, for example `fun \unsafe-1\ foo()`.
```

Comment:
Would these annotations apply transitively?

That is, if `foo` is marked with `unsafe-1`, am I required to put the same mark on any function that calls `foo`?

Reply:
No, these annotations aren't transitive for the reasons I explained in my reply to your other comment.